Make FFT algorithms public

ejmahler opened the pull request:
This PR makes changes 1 and 3 mentioned in #13 

- It renames DFTAlgorithm to just DFT, and replaces the dft() free function with the DFT struct
- It makes all FFT algorithms public, by simply making the algorithm module public
- It performs a few small cleanup operations to make sure the algorithms are safe for public use

Commit:
Replace uses of NoopAlgorithm with DFTAlgorithm

Commit:
Replaced dft() free function with DFT struct

Commit:
Made algorithm module public

Commit:
Reduced memory footprint of ButterflyEnum

Commit:
Removed unsafe blocks and added unit test for GoodThomasAlgorithm

Commit:
Improved test coverage for all algorithms

awelkie commented on src/algorithm/mod.rs:
```diff
@@ -31,9 +32,9 @@ pub enum ButterflyEnum<T> {
 	Butterfly3(butterflies::Butterfly3<T>),
 	Butterfly4(butterflies::Butterfly4),
-	Butterfly5(butterflies::Butterfly5<T>),
+	Butterfly5(Box<butterflies::Butterfly5<T>>),
```
Why are `Butterfly5` and `Butterfly7` boxed now? Is this just to save space on the enum size?

ejmahler replied:
Yes. If you look at the size of the MixedRadixDoubleButterfly, right now it's over 100 bytes for f32, and over 200 bytes for f64, and almost all of it comes from the 6-element array in the butterfly7 struct.

For most FFTs this will be purely wasted space. I think we have to be ok with some wasted space, but 5 and 7 are the two worst for space.

awelkie commented:
I just took a look at the API from `cargo doc` and it looks good. I'll be able to take a look into the code changes more tomorrow.

Commit:
Refactored FFTButterfly trait and tests

ejmahler commented:
Pushed one final commit for this PR, and it's ready to go. I've been wanting to refactor some elements of the butterflies ever since I wrote them, and I just now realized that if they're going to be public, then I need to do it now.

Basically, there were a few problems before:
- For several of the butterflies, it was difficult to call len() on them, because the struct doesn't have a type parameter, but the FFTAlgorithm trait which owns len() does. there are many places in each butterfly impl where we have to hard-code the butterfly size, and it would be much more maintainable to be able to call self.len() instead. So I moved len() into its own trait that doesn't hav a type parameter, and made FFTAlgorithm and FFTButterfly depend on it. It looks very nice this way imo.
- In its previous state, it was awkward to deal with instances of ButterflyEnum. I gave it a new() method and made it implement FFTbutterfly, since it was halfway to implementing it anyways.
- The different butterflies handled the "perform_FFT" method inconsistently, and as a part of making it more consistent i added it to the FFTButterfly trait as "process_inplace"
- There was a **ton** of code duplication for unit tests, and the above changes allowed me to get rid of almost all of it.

Basically -- make burtterflies more consistent and ergonomic, and reduce code duplication

ejmahler commented:
And most importantly, this last PR means that every single public method has test coverage!

awelkie commented:
Looks good to me. The only thing I don't really like is the inconsistency in the boxing of the `ButterflyEnum`. I understand the reason for it, but it just makes for a sort of weird API.

What do you think about instead of `ButterflyEnum`, we just hold trait objects of `FFTButterfly`? This seems like a cleaner API to me, and I don't see any downsides as far as runtime computation.

Also, I would like the algorithms documented before pushing to crates.io. But I can do this after the PR is merged, and indeed I would like to because it would give me a good chance to solidify my knowledge of each of the algorithms.

Commit:
Make MixedRadixDoubleButterfly operate on trait objects

ejmahler commented:
Based on your feedback, I modified MixedRadixDoubleButterfly to operate on FFTButterfly trait objects, and deleted ButterflyEnum entirely. It looks much nicer this way, it's more consistent with the other FFT algorithms, and it had no effect on any of the benchmarks.

ejmahler commented:
The entire reason ButterflyEnum existed in the first place is that I assumed storing the butterfly inline with the MixedRadixDoubleButterfly struct would improve performance. But it looks like I was wrong! So we don't need it at all.

awelkie commented on src/algorithm/mixed_radix.rs:
```diff
@@ -93,16 +97,16 @@ impl<T: FFTnum> FFTAlgorithm<T> for MixedRadix<T> {
 pub struct MixedRadixDoubleButterfly<T> {
     width: usize,
-    width_size_fft: ButterflyEnum<T>,
+    width_size_fft: Rc<FFTButterfly<T>>,
```
Why use `Rc` here instead of `Box`?

ejmahler replied:
For the same reason that MixedRadix takes a Rc instead of a box: It's so that the planner can cache instances of it. Check out plan.rs line 42, where it checks if we already have an instance of that butterfly before making a new one.

I'm trying to think of reasons why it should be Box. One that I can come up with is lower memory overhead if butterfly instances aren't frequently re-used. what do you think?

awelkie replied:
Ah, no that makes sense. I think `Rc` is fine here, I was just curious because I didn't see the caching.
